Jakogut/crun

jakogut opened the pull request:
This runtime is smaller, and allows for faster container startup and
    lower resource usage compared to runc.
    
    The runtime can either be specified on the command line:
    
    $ balena run --runtime=crun hello-world
    
    Or as the default runtime in the engine's daemon.json config:
    
    $ cat << EOF > /etc/docker/daemon.json
    {
      "runtimes": {
        "crun": {"path": "/usr/bin/crun"}
      },
      "default-runtime": "crun"
    }
    EOF

jakogut commented:
Related to: https://github.com/balena-os/meta-balena/issues/2326

jakogut commented:
@balena-ci rebase

alexgg commented on meta-balena-common/recipes-security/libseccomp/files/run-ptest:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+cd tests
+./regression -a
```
balenaOS does not include the ptest distro feature, so this is not needed.

alexgg commented on meta-balena-common/recipes-containers/oci-runtime-spec/oci-runtime-spec_git.bb:
```diff
@@ -0,0 +1,41 @@
+
+FILES_${PN} += "${prefix}/local/go/src/${PKG_NAME}/*"
+
+CLEANBROKEN = "1"
```
What is this recipe installing exactly, and does it really need to go into the device's filesystem or is it only a compile time dependency?

jakogut replied:
Just a source file, as indicated by the recipe

alexgg commented on meta-balena-common/recipes-containers/oci-runtime-spec/oci-runtime-spec_git.bb:
```diff
@@ -0,0 +1,41 @@
+
+FILES:${PN} = "${prefix}/local/go/src/${PKG_NAME}/*"
+
+CLEANBROKEN = "1"
```
What is this recipe installing exactly, and does it really need to go into the device's filesystem or is it only a compile time dependency?

alexgg commented on meta-balena-common/recipes-containers/oci-image-spec/oci-image-spec_git.bb:
```diff
@@ -0,0 +1,41 @@
+
+FILES:${PN} = "${prefix}/local/go/src/${PKG_NAME}/*"
+
+CLEANBROKEN = "1"
```
What is this recipe installing exactly, and does it really need to go into the device's filesystem or is it only a compile time dependency?

alexgg commented on meta-balena-common/recipes-core/packagegroups/packagegroup-resin.bb:
```diff
@@ -17,6 +17,7 @@ RDEPENDS:${PN} += " \
     mobynit \
     docker-disk \
+    crun \
     hostapp-update \
```
Given that this is added unconditionally, what is the rootfs size increase?

jakogut replied:
@alexgg Looks like about 476K increase, based on the buildhistory diff

```
-IMAGESIZE = 675296
+IMAGESIZE = 675772
```

Additionally, it appears the source files mentioned below aren't included in the image after all. Here's the full output from the buildhistory diff: https://gist.github.com/jakogut/dd575d995afddfde90fc0e29f2e109b2

alexgg commented on meta-balena-common/recipes-security/libseccomp/libseccomp_2.5.1.bb:
```diff
@@ -0,0 +1,47 @@
+    for file in $(find tools/* -executable -type f); do
+        install -m 744 ${S}/${file} ${D}/${PTEST_PATH}/tools
+    done
+}
```
Please remove all the ptest related code, better on a different commit so that it's clear what is the original recipe and it's source, and then the balenaOS specific modifications.

alexgg commented on meta-balena-common/recipes-security/libseccomp/libseccomp_2.5.1.bb:
```diff
@@ -0,0 +1,47 @@
+FILES:${PN} = "${bindir} ${libdir}/${BPN}.so*"
+FILES:${PN}-dbg += "${libdir}/${PN}/tests/.debug/* ${libdir}/${PN}/tools/.debug"
+
+RDEPENDS:${PN}-ptest = "coreutils bash"
```
As above, remove ptest related code.

alexgg commented on meta-resin-pyro/recipes-containers/crun/crun_%.bbappend:
```diff
@@ -0,0 +1 @@
+PACKAGECONFIG_remove = "systemd"
```
This doesn't seem right as balenaOS is running systemd. What's the reason for this?

alexgg replied:
I seem to remember older Yocto releases have a systemd version that does not work with crun, so DTs with older Yocto releases cannot use crun.
It maybe better just to append the packagegroup in these older layers to remove crun.

jakogut replied:
@alexgg Crun uses systemd for certain features, but expects certain symbols in libsystemd that are not present in the versions provided by pyro and earlier releases. That doesn't mean the runtime doesn't work (though it remains to be tested), but that the features provided by linking to systemd are disabled.

Glancing at the man page and code, it appears that this is primarily used for configuring cgroups through systemd, though cgroups can also be configured directly through the cgroupfs backend.

alexgg replied:
@jakogut unless we can guarantee crun works fine, I'd rather not include it. So it's either preparing and running a comprehensive test plan for older releases or not including it in them.

jakogut replied:
@alexgg What would a comprehensive test plan involve? I've already added a test case for ensuring that crun can be configured as the default runtime, and can run containers on a given release.

alexgg replied:
@jakogut I personally don't like the idea of diverging the OS builds with two different runtimes, so what I am thinking about is in using the best runtime. If that is crun, then let's make it the default.
With that in mind, we are speaking about replacing the container runtime, so it can potentially affects all containers, from the supervisor to user apps. A test plan would need to give us 100% guarantee that there are no unknown problems lurking around. Something like choosing a representative variety of applications and making sure they work on all the different architectures supported, adding some stress testing with a couple of apps, long run testing to make sure it doesn't have long term effects (memory leaks etc), and then carefully staging a release on smaller to bigger fleets to monitor and fix any problem. Something at least as comprehensive as we have done for the overlay migration.
Probably a good idea to merge this with an engine & go update, so that the test plan can be shared.
That's my point of view which I acknowledge tends to be conservative - once we speak about it on an arch call we will have a broader variety of opinions.

alexgg replied:
@jakogut I didn't quite answer the question though - the above is the test plan we would need to make crun the default runtime. Then we need to either repeat it for older releases were we build it differently, or just not replace it on older Yocto releases (which I prefer)

alexgg commented on meta-resin-rocko/recipes-containers/crun/crun_%.bbappend:
```diff
@@ -0,0 +1 @@
+PACKAGECONFIG_remove = "systemd"
```
As above

alexgg commented on meta-resin-sumo/recipes-containers/crun/crun_%.bbappend:
```diff
@@ -0,0 +1 @@
+PACKAGECONFIG_remove = "systemd"
```
As above

alexgg reviewed (changes requested):
There are some comments and open questions, we also need a list of advantages for the scheduled arch call. The decision to make is whether we want to make it default and what is the amount of testing needed to make the swap with confidence.

jakogut commented:
@alexgg Following up on your review comment:

> we also need a list of advantages for the scheduled arch call. The decision to make is whether we want to make it default and what is the amount of testing needed to make the swap with confidence.

I've added a balena-config-vars script and service to configure the engine with a default container runtime defined in `config.json`, which allows us to continue using runc, but switch to crun (or any other OCI compatible runtime) at runtime, rather than defining this universally in the build. Assuming the slightly increased image size doesn't push us over the edge on any device types, this PR should allow us to experiment with the new runtime with minimal risk.

As far as the advantages go, [this](https://www.redhat.com/sysadmin/introduction-crun) article from RedHat explains them pretty well.

However, I'll also summarize:
- crun binary is ~30-50x smaller than standalone runc, and ~100x smaller than the balena-engine binary we symlink to `/usr/bin/balena-runc` **
- crun consumes less memory than runc (~3.7M compared to ~15M), and uses only one thread, allowing containers to run in more resource constrained environments, such as the Pi Zero. RedHat claims to have experimented with running containers with memory limits as low as 250K.
- crun speeds up container startup time, up to 40% in my testing on the Pi Zero on balenaOS.
- crun supports experimental features, before they've been standardized and implemented in runc
- crun is more portable, allowing Docker to run on architectures where Go support is limited, this is how Docker was initially ported to RISC-V

** The runc runtime is compiled into the engine binary, and removing it would save a negligible amount of space. However, a smaller container runtime allows fewer page faults on container startup, which means fewer reads from disk, less memory usage, and faster container startup.

alexgg commented:
> this PR should allow us to experiment with the new runtime with minimal risk.

This is fine for a development branch, but no experiments in production. And we also don't want to have to start a support conversation with "what container runtime are you using?".

So the way forward as I see is agree that is is a better runtime, test it thoroughly and make it the default.

jakogut commented:
> we also don't want to have to start a support conversation with "what container runtime are you using?"

What about runtimes that are OCI-compliant, but purposefully implement different behaviors, such as [katacontainers](https://katacontainers.io/) which run workloads in lightweight virtual machines?

alexgg commented:
> What about runtimes that are OCI-compliant, but purposefully implement different behaviors,

I'd argue the same apply, if it's a better choice let's make it the default. I don't see advantages in a runtime switch, and it makes the product more complex. I don't think customers should care what runtime their container is running on. It's a product/arch call question anyway, so bring it forward if you think otherwise.

klutchell commented:
@resin-jenkins retest this please
